elastic/elastic-agent
# Use package manifests for upgrade

## What does this PR do?
This is a subset of a bigger PR #3950, so CI is not expected to be green, especially regarding the integration tests.

This PR uses the package manifest included in elastic-agent packages to map the package contents onto disk when upgrading . In the context of #2579 the path mapping is used to install the agent we are upgrading to in a directory like `data/elastic-agent-<package version>-<hash>`.

The upgrade process now looks for the package manifest file and copies the files in the archive after mapping the name using the path mappings in the manifest. This means that also the new agent will end up in `data/elastic-agent-<package version>-<hash>`



## Why is it important?



This PR is needed to ensure that new versions of agents get installed to a new path containing the version from a package `data/elastic-agent-<package version>-<hash>` that still uses the legacy structure `data/elastic-agent-<hash>` for backward compatibility. In practice it allows for upgrading to an agent version built from the same commit as the existing agent as long as the package version is different (the goal of #2579)

## Checklist




## Author's Checklist



## How to test this PR locally


To test this PR locally, we package elastic-agent twice from the same commit specifying a different AGENT_PACKAGE_VERSION like:
```
EXTERNAL=true SNAPSHOT=true PLATFORMS="linux/amd64" PACKAGES="tar.gz"  mage package
```
and
```
AGENT_PACKAGE_VERSION=8.13.0-repackaged EXTERNAL=true SNAPSHOT=true PLATFORMS="linux/amd64" PACKAGES="tar.gz"  mage package
```

Install the first agent package and verify that it's healthy and it runs from `data/elastic-agent-8.13.0-SNAPSHOT-<hash>` directory.
Upgrade this agent using the second package
```
elastic-agent upgrade 8.13.0-repackaged-SNAPSHOT --source-uri file://<directory where the package is located on disk> --skip-verify
```

Upgrade will unpack and store the new agent in `data/elastic-agent-8.13.0-repackaged-<hash>` and rotate the links.
Verify the agent directory structure contains the 2 distinct directories (you have to do it before the grace period ends).

*Known issues*:
- When the grace period expires the watcher will break the agent install since it will clean anything that does not look like `elastic-agent-<hash>`, this is addressed already in a follow-up PR
- An extra `elastic-agent-<hash>` directory is created during upgrade due to using the wrong path during copy of component states. This is addressed in a follow-up PR
- This has been tested only on Linux. MacOS and Windows fixes will come in a follow-up PR (we need small modifications to make this work mostly related to OS-dependent agent paths).
- To clean up the broken agent using uninstall (it can be done manually as well but I find this more convenient):
  1. remove the broken symlink in `/opt/Elastic/Agent`
  2. use the extracted package to force  install the agent again `./elastic-agent install --force`
  3. remove using uninstall `sudo elastic-agent uninstall`


## Related issues

- Relates #2579 
- Relates #3950 
- Requires #4173



- 

## Use cases



## Screenshots



## Logs



## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...
